### ISSUE TITLE
feat: profiling?

### ISSUE THREAD

**Issue by NickCrews** [ibis-project/ibis#9276]
### Is your feature request related to a problem?

This is a bit broad, IDK if this should be a discussion.

I want to be able to profile queries. Specifically, which parts of the query are slow. This is tricky, because in ibis computation happens in two stages:

1. expression building: This is always fast, since we aren't actually processing any data. But this is where I want to be able to assign blame, ie the `table.distinct(on="my_id")` call is responsible for 90% of the query runtime. Using timeit or other python profiling is useless here.
2. runtime: This is where the big data is actually processed, and where the time actually happens.

I can currently do A/B testing, by just timing the whole run, but this is annoying to have to think about what to change, and then writing the change.

### What is the motivation behind your request?

I am motived both as an end user, but also as [a library author](https://github.com/NickCrews/mismo/issues/42).

### Describe the solution you'd like

Ideally every Operation could get associated with a run time, so for instance if you plot an expression graph you could see the runtime next to each node.

That seems very difficult though. Still very usable would be if we just got back the raw, non-machine readable output from eg [`EXPLAIN ANALYZE <QUERY>`](https://duckdb.org/docs/dev/profiling.html#run-time-profiling-with-the-explain-analyze-statement). This solution I don't think would even need to be implemented in ibis, but it would be ergonomic if it was a simple `expr.profile()` or `ibis.profile(expr)`.

I'm not sure about backend support. I don't know, but I imagine most of the SQL backends support some sort of profiling. [SQlite](https://www.sqlite.org/profile.html) profiling looks very different from duckdb profiling.

### What version of ibis are you running?

main

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by cpcloud** [comment]
Hey Nick!

This is a super interesting problem, but I think it's not feasible for us to correlate an ibis expression with a particular duration (or some proxy of that like instructions, or even database plan node) because the thing that the database executes often looks very different from the thing it was asked to execute.

It's one of the trade-offs associated with using SQL and/or a database: you generally get better performance at the cost of debugability.

That said, what do you think about a `backend.explain(expr)` method? We had this a while ago but it was removed due to lack of implementation across our backends.

This would at least allow you to get some insight into what the database thinks is happening.

**Comment by NickCrews** [comment]
>  what do you think about a backend.explain(expr) method?

I think that sounds nice, even if it returns teh raw string. However, to avoid needing to do a private method on a non-explicit-backend expression

```python
e = ibis.read_csv().<sdfgdsfg>
e._find_backend().explain(e)
```

I would love to either (1 is my favorite)
1. have `ibis.explain(expr)` in addition to `backend.explain(expr)`
1. just have `expr.explain()` (though we don't want to fill up that namespace TOO much)
1. have `expr.explain()` in addition to `backend.explain(expr)`

Possible things to support ([duckdb for inspiration](https://duckdb.org/docs/guides/meta/explain.html))
- (need) show physical plan
- (need) show profiling/timing results
- (nice to have) show logical plan
- (nice to have) show optimized logical plan

**Comment by cboettig** [comment]
Would it be possible to support the duckdb explain mechanism (i.e. have an `ibis.duckdb.explain()` at least as an intermediate?  Maybe this already exists (or there's some way to trigger this), but I'm only seeing a `.explain` method for the impala backend I think?

**Comment by dangotbanned** [comment]
My motivation is pretty different (*typing*), but I was hoping to find an `ibis.Table.explain` method.


### Survey
In [`narwhals`](https://github.com/narwhals-dev/narwhals), the equivalent to `ibis.Table` in all *other* lazy backends we support has an `explain` method:
- [`polars.LazyFrame.explain`](https://docs.pola.rs/api/python/stable/reference/lazyframe/api/polars.LazyFrame.explain.html)
- [`duckdb.DuckDBPyRelation.explain`](https://duckdb.org/docs/stable/clients/python/relational_api#explain)
- [`pyspark.sql.DataFrame.explain`](https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.DataFrame.explain.html#pyspark.sql.DataFrame.explain)
- [`sqlframe.base.dataframe.DataFrame.explain`](https://github.com/eakmanrq/sqlframe/blob/961d3fda2fb7a28385a01b0b3f4f49dbe4c1e7d2/sqlframe/base/dataframe.py#L1301-L1370)
- [`dask.dataframe.DataFrame.explain`](https://docs.dask.org/en/stable/generated/dask.dataframe.DataFrame.explain.html#dask.dataframe.DataFrame.explain)
- `ibis.Table.explain` 😔

Additionally, a [user-requested backend](https://github.com/narwhals-dev/narwhals/issues/2222) `daft`:
- [`daft.DataFrame.explain`](https://docs.daft.ai/en/stable/api/dataframe/#daft.DataFrame.explain)

**I thought this detail might be more relevant to `ibis` than my use-case**

---

### Background
> [!NOTE]
> **TL;DR:** `explain` would help us avoid depending on `ibis` internals (for typing)

Downstream, I've been working on some typing which can be used for matching objects from optional dependencies (like `ibis.Table`) in [`@overload` definitions](https://typing.python.org/en/latest/spec/overload.html):

<details><summary><b>Show related PRs</b></summary>
<p>

- https://github.com/narwhals-dev/narwhals/pull/2159
- https://github.com/narwhals-dev/narwhals/pull/2356
- https://github.com/narwhals-dev/narwhals/pull/2324
- https://github.com/narwhals-dev/narwhals/pull/2944
- https://github.com/narwhals-dev/narwhals/pull/3016
- https://github.com/narwhals-dev/narwhals/pull/3086
- https://github.com/narwhals-dev/narwhals/pull/3125

</p>
</details>

That's a lot of content 😅, but here's a more condensed repro of the original problem:

<details><summary><b>Show repro </b>(problem)</summary>
<p>

```py
from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal, overload

from typing_extensions import reveal_type

if TYPE_CHECKING:
    import collections

    import cudf
    import dask.dataframe as dd
    import ibis
    import pandas as pd
    import polars as pl
    import pyspark.sql
    from typing_extensions import TypeAlias, TypeIs

Known: TypeAlias = Literal["polars", "dask", "pandas", "ibis", "pyspark", "cudf"]

KNOWN_PACKAGES: tuple[Known, ...] = (
    "polars",
    "dask",
    "pandas",
    "ibis",
    "pyspark",
    "cudf",
)


def is_known(package_name: str) -> TypeIs[Known]:
    return package_name in KNOWN_PACKAGES


@overload
def package(native: pl.DataFrame) -> Literal["polars"]: ...
@overload
def package(native: pd.DataFrame) -> Literal["pandas"]: ...
@overload
def package(native: ibis.Table) -> Literal["ibis"]: ...  # mypy-only issue (1)
@overload
def package(native: cudf.DataFrame) -> Literal["cudf"]: ...  # not installed (1)
@overload
def package(native: dd.DataFrame) -> Literal["dask"]: ...  # mypy-only issue (2)
@overload
def package(native: pyspark.sql.DataFrame) -> Literal["pyspark"]: ...  # not installed (2)
@overload
def package(native: Any) -> Literal["unknown"]: ...


def package(native: Any) -> Literal[Known, "unknown"]:
    package_name = type(native).__module__.split(".")[0]
    if is_known(package_name):
        return package_name
    return "unknown"


def test_package(
    a: pl.DataFrame,
    b: pd.DataFrame,
    c: ibis.Table,
    d: cudf.DataFrame,
    e: dd.DataFrame,
    f: pyspark.sql.DataFrame,
    g: collections.deque[Any],
) -> None:
    r_polars = package(a)
    r_pandas = package(b)
    r_ibis = package(c)
    r_cudf = package(d)
    r_dask = package(e)
    r_pyspark = package(f)
    r_unknown = package(g)

    reveal_type(r_polars)
    reveal_type(r_pandas)
    reveal_type(r_ibis)
    reveal_type(r_cudf)
    reveal_type(r_dask)
    reveal_type(r_pyspark)
    reveal_type(r_unknown)

```

</p>
</details>

These outputs are fully-dependent on which packages are installed. In my case, everything but `pyspark` and `cudf`.
This alone is enough to break the typing for anything beyond the first `@overload` mentioning either package:


<details><summary><b>Show output</b> <code>pyright</code> (problem)</summary>
<p>

```py
pyright narwhals/t.py
narwhals/t.py
  narwhals/t.py:74:17 - information: Type of "r_polars" is "Literal['polars']"
  narwhals/t.py:75:17 - information: Type of "r_pandas" is "Literal['pandas']"
  narwhals/t.py:76:17 - information: Type of "r_ibis" is "Literal['ibis']"
  narwhals/t.py:77:17 - information: Type of "r_cudf" is "Unknown"
  narwhals/t.py:78:17 - information: Type of "r_dask" is "Literal['cudf']"
  narwhals/t.py:79:17 - information: Type of "r_pyspark" is "Unknown"
  narwhals/t.py:80:17 - information: Type of "r_unknown" is "Literal['cudf']"
0 errors, 0 warnings, 7 informations
```

</p>
</details>

`mypy` further complicates things, for the reasons discussed in (https://github.com/ibis-project/ibis/issues/6844):

<details><summary><b>Show output</b> <code>mypy</code> (problem)</summary>
<p>

```py
mypy narwhals/t.py
narwhals/t.py:41: error: Overloaded function signature 4 will never be matched: signature 3's parameter type(s) are the same or broader  [overload-cannot-match]
    def package(native: cudf.DataFrame) -> Literal["cudf"]: ...  # not installed (1)
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
narwhals/t.py:43: error: Overloaded function signature 5 will never be matched: signature 3's parameter type(s) are the same or broader  [overload-cannot-match]
    def package(native: dd.DataFrame) -> Literal["dask"]: ...  # mypy-only issue (2)
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
narwhals/t.py:43: error: Overloaded function signature 5 will never be matched: signature 4's parameter type(s) are the same or broader  [overload-cannot-match]
    def package(native: dd.DataFrame) -> Literal["dask"]: ...  # mypy-only issue (2)
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
narwhals/t.py:45: error: Overloaded function signature 6 will never be matched: signature 3's parameter type(s) are the same or broader  [overload-cannot-match]
    def package(native: pyspark.sql.DataFrame) -> Literal["pyspark"]: ...  # not installed (2)
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
narwhals/t.py:45: error: Overloaded function signature 6 will never be matched: signature 4's parameter type(s) are the same or broader  [overload-cannot-match]
    def package(native: pyspark.sql.DataFrame) -> Literal["pyspark"]: ...  # not installed (2)
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
narwhals/t.py:45: error: Overloaded function signature 6 will never be matched: signature 5's parameter type(s) are the same or broader  [overload-cannot-match]
    def package(native: pyspark.sql.DataFrame) -> Literal["pyspark"]: ...  # not installed (2)
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
narwhals/t.py:47: error: Overloaded function signature 7 will never be matched: signature 3's parameter type(s) are the same or broader  [overload-cannot-match]
    def package(native: Any) -> Literal["unknown"]: ...
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
narwhals/t.py:47: error: Overloaded function signature 7 will never be matched: signature 4's parameter type(s) are the same or broader  [overload-cannot-match]
    def package(native: Any) -> Literal["unknown"]: ...
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
narwhals/t.py:47: error: Overloaded function signature 7 will never be matched: signature 5's parameter type(s) are the same or broader  [overload-cannot-match]
    def package(native: Any) -> Literal["unknown"]: ...
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
narwhals/t.py:47: error: Overloaded function signature 7 will never be matched: signature 6's parameter type(s) are the same or broader  [overload-cannot-match]
    def package(native: Any) -> Literal["unknown"]: ...
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
narwhals/t.py:74: note: Revealed type is "Literal['polars']"
narwhals/t.py:75: note: Revealed type is "Literal['pandas']"
narwhals/t.py:76: note: Revealed type is "Any"
narwhals/t.py:77: note: Revealed type is "Any"
narwhals/t.py:78: note: Revealed type is "Any"
narwhals/t.py:79: note: Revealed type is "Any"
narwhals/t.py:80: note: Revealed type is "Literal['ibis']"
Found 10 errors in 1 file (checked 1 source file)
```

</p>
</details>

#### Current solution
Everything besides `ibis.Table` can match to [`NativeLazyFrame`](https://github.com/narwhals-dev/narwhals/blob/3040310a1feb022a3ec1f82d01ef8ae30cdc5246/narwhals/_native.py#L155-L168) due to the presence of `explain`.
Then defining a special [protocol for `ibis.Table`](https://github.com/narwhals-dev/narwhals/blob/3040310a1feb022a3ec1f82d01ef8ae30cdc5246/narwhals/_native.py#L214-L218) and using a [`Union` w/ what works for the others](https://github.com/narwhals-dev/narwhals/blob/3040310a1feb022a3ec1f82d01ef8ae30cdc5246/narwhals/_native.py#L263).

The positive cases here work fine:

<details><summary><b>Show repro</b> (solution)</summary>
<p>

```py
from __future__ import annotations

from typing import TYPE_CHECKING

from typing_extensions import reveal_type

import narwhals as nw

if TYPE_CHECKING:
    import ibis
    import pandas as pd
    import polars as pl

    from narwhals.typing import IntoLazyFrameT


def from_lazy(native: IntoLazyFrameT) -> nw.LazyFrame[IntoLazyFrameT]:
    return nw.from_native(native)


def example(good_1: ibis.Table, good_2: pl.LazyFrame, bad: pd.DataFrame) -> None:
    r_good_1 = from_lazy(good_1)
    r_good_2 = from_lazy(good_2)
    r_bad = from_lazy(bad)

    reveal_type(r_good_1)
    reveal_type(r_good_2)
    reveal_type(r_bad)

```

</p>
</details>

But the negative cases (e.g. `pd.DataFrame` isn't lazy) leak this workaround out into the diagnostic messages:


<details><summary><b>Show output</b> <code>pyright</code> (solution)</summary>
<p>

```py
pyright narwhals/t.py
narwhals/t.py
  narwhals/t.py:24:23 - error: Argument of type "DataFrame" cannot be assigned to parameter "native" of type "IntoLazyFrameT@from_lazy" in function "from_lazy"
    Type "DataFrame" is not assignable to type "IntoLazyFrame"
      Type "DataFrame" is not assignable to type "IntoLazyFrame"
        "DataFrame" is incompatible with protocol "NativeLazyFrame"
          "explain" is not present
            "NativeFrame" is not assignable to "DataFrame"
        "DataFrame" is incompatible with protocol "_NativeIbis"
          "sql" is not present
          "__pyarrow_result__" is not present
    ... (reportArgumentType)
  narwhals/t.py:26:17 - information: Type of "r_good_1" is "LazyFrame[Table]"
  narwhals/t.py:27:17 - information: Type of "r_good_2" is "narwhals.dataframe.LazyFrame[polars.lazyframe.frame.LazyFrame]"
  narwhals/t.py:28:17 - information: Type of "r_bad" is "LazyFrame[Unknown]"
1 error, 0 warnings, 3 informations
```

</p>
</details>
